src/kudu/consensus/consensus_peers.cc

```diff
@@ -206,8 +210,28 @@ Status Peer::SignalRequest(bool even_if_queue_empty) {
   return Status::OK();
 }
 
+bool Peer::ProxyBatchDurationHasPassed() {
+  const std::string& uuid = peer_pb_.permanent_uuid();
+  std::string next_hop_uuid;
+
+  // TODO: The routing table is consulted again in RequestForPeer(), ideally we
+  // can do it once
+  queue_->GetNextRoutingHopFromLeader(uuid, &next_hop_uuid);
+
+  const bool peer_is_proxied = next_hop_uuid != uuid;
+
+  return !peer_is_proxied ||
+      MonoTime::Now() - last_request_time_ >=
+       MonoDelta::FromMilliseconds(FLAGS_proxy_batch_duration_ms);
+}
+
 void Peer::SendNextRequest(bool even_if_queue_empty) {
   std::unique_lock<simple_spinlock> l(peer_lock_);
+
+  if (!ProxyBatchDurationHasPassed()) {
```

Comment:
This will also block out HBs if proxy delay > HB interval. I think the only way to avoid filtering out HBs is by making sure that the proxy delay is always <= HB interval. Still looking into this.